bpf/lib/lb.h

```diff
@@ -112,55 +127,77 @@ static inline int __inline__ reverse_map_l4_port(struct __sk_buff *skb, __u8 nex
 	return 0;
 }
 
-/** Perform IPv6 DSR SNAT based on reverse NAT index
- * @arg skb		packet
- * @arg l4_off		offset to L4
- * @arg csum_off	offset to L4 checksum field
- * @arg csum_flags	checksum flags
- * @arg index		reverse NAT index
- * @arg tuple		tuple
- */
-static inline int lb6_dsr_snat(struct __sk_buff *skb, int l4_off, int csum_off,
-			       int csum_flags, __u16 index, struct ipv6_ct_tuple *tuple)
+static inline int __inline__ __lb6_rev_nat(struct __sk_buff *skb, int l4_off,
+					 struct csum_offset *csum_off,
+					 struct ipv6_ct_tuple *tuple, int flags,
+					 struct lb6_reverse_nat *nat)
 {
-	union v6addr tmp, sip;
-	struct lb6_reverse_nat *nat;
+	union v6addr old_saddr;
+	union v6addr tmp;
+	__u8 *new_saddr;
 	__be32 sum;
 	int ret;
 
-	cilium_trace(skb, DBG_LB6_REVERSE_NAT_LOOKUP, index, 0);
-	nat = map_lookup_elem(&cilium_lb6_reverse_nat, &index);
-	if (nat == NULL) {
-		return 0;
+	cilium_trace(skb, DBG_LB6_REVERSE_NAT, nat->address.p4, nat->port);
+
+	if (nat->port) {
+		ret = reverse_map_l4_port(skb, tuple->nexthdr, nat->port, l4_off, csum_off);
+		if (IS_ERR(ret))
+			return ret;
 	}
 
-	cilium_trace(skb, DBG_LB6_REVERSE_NAT, nat->address.p4, nat->port);
-	ret = reverse_map_l4_port(skb, tuple->nexthdr, nat->port, l4_off, csum_off);
-	if (IS_ERR(ret))
-		return ret;
+	if (flags & REV_NAT_F_TUPLE_SADDR) {
+		ipv6_addr_copy(&old_saddr, &tuple->addr);
+		ipv6_addr_copy(&tuple->addr, &nat->address);
+		new_saddr = tuple->addr.addr;
```

Comment:
What I meant was the logic would work if old_saddr is the same as what would be obtained via ipv6_load_saddr(skb, ETH_HLEN, &old_saddr) so that a later csum_l4_replace will work fine. Because the skb csum was originally computed using whatever address was in the skb ? And the replace is simply offsetting the diff and not recomputing the entire csum.